In [99]:
import pandas as pd
import numpy as np

Loading the Dataset

In [100]:
df = pd.read_csv("/Users/mdnaif/Desktop/E2E Project Practice/Telco-Churn-Prediction/data/WA_Fn-UseC_-Telco-Customer-Churn.csv")

Understanding the Dataset

In [101]:
print(f" Shape: {df.shape} \n Size: {df.size}")

 Shape: (7043, 21) 
 Size: 147903


as we can observe from above that there are :
- 7043 Rows & 21 Columns
- The dataset is of 147903 in Size

In [102]:
df.head(10)

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes
5,9305-CDSKC,Female,0,No,No,8,Yes,Yes,Fiber optic,No,...,Yes,No,Yes,Yes,Month-to-month,Yes,Electronic check,99.65,820.5,Yes
6,1452-KIOVK,Male,0,No,Yes,22,Yes,Yes,Fiber optic,No,...,No,No,Yes,No,Month-to-month,Yes,Credit card (automatic),89.10,1949.4,No
7,6713-OKOMC,Female,0,No,No,10,No,No phone service,DSL,Yes,...,No,No,No,No,Month-to-month,No,Mailed check,29.75,301.9,No
8,7892-POOKP,Female,0,Yes,No,28,Yes,Yes,Fiber optic,No,...,Yes,Yes,Yes,Yes,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes
9,6388-TABGU,Male,0,No,Yes,62,Yes,No,DSL,Yes,...,No,No,No,No,One year,No,Bank transfer (automatic),56.15,3487.95,No


- customerID is of no use to us so we will drop it.
- extensive useage of encoding is required as most columns seems to be object type.
- it is a binary classifcation problem statement as there are 2 outputs in the Churn Column.

In [103]:
df.drop(columns="customerID",inplace=True)

In [104]:
df.columns

Index(['gender', 'SeniorCitizen', 'Partner', 'Dependents', 'tenure',
       'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity',
       'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV',
       'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod',
       'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

In [105]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 20 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   gender            7043 non-null   object 
 1   SeniorCitizen     7043 non-null   int64  
 2   Partner           7043 non-null   object 
 3   Dependents        7043 non-null   object 
 4   tenure            7043 non-null   int64  
 5   PhoneService      7043 non-null   object 
 6   MultipleLines     7043 non-null   object 
 7   InternetService   7043 non-null   object 
 8   OnlineSecurity    7043 non-null   object 
 9   OnlineBackup      7043 non-null   object 
 10  DeviceProtection  7043 non-null   object 
 11  TechSupport       7043 non-null   object 
 12  StreamingTV       7043 non-null   object 
 13  StreamingMovies   7043 non-null   object 
 14  Contract          7043 non-null   object 
 15  PaperlessBilling  7043 non-null   object 
 16  PaymentMethod     7043 non-null   object 


- there does'nt seem to be any null value but we will crosscheck just in case.
- majority of columns are object data type.
- only 3 columns are numerical.

In [106]:
df.describe()

,SeniorCitizen,tenure,MonthlyCharges
count,7043.000000,7043.000000,7043.000000
mean,0.162147,32.371149,64.761692
std,0.368612,24.559481,30.090047
min,0.000000,0.000000,18.250000
25%,0.000000,9.000000,35.500000
50%,0.000000,29.000000,70.350000
75%,0.000000,55.000000,89.850000
max,1.000000,72.000000,118.750000


we can observe from the above dataset that:
- Senior Citizen is already binary encoded : 0 & 1
- tenure : min value = 0.00 month, max value = 72 months
- monthly charges : min value = 18.25, max value = 118.75

as observed from above we can say that customers tenure ranged between 0 to 6 years, which gives us the insight that  few customers left early without even staying for a full month and the average customers stayed for 2 years 8 month while some stayed loyal and conntined the service for 6 years. The monthy charges ranged between $18.25 the cheapest package and some opting for premium packages at $118.75 with the average customer churn price being $64.76. 

In [107]:
df.isna().sum()

gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

- dataset is clear of null values.

In [108]:
print(f"Duplicated Rows : {df.duplicated().sum()}")

Duplicated Rows : 22


- without dropping the customerID it showed 0 duplicates
- after droping the customerID it showed 22 duplicates

what this means is that with customerID we were able to uniquely identify customers, since we dropped it we seem to observe 22 duplicate rows which represents simmilar customer behaviour so we don't need to worry about it.

In [109]:
df["Churn"].value_counts()

Churn
No     5174
Yes    1869
Name: count, dtype: int64

- the dataset is highly imbalanced.
- we can observe that: customers that stayed > customer that left
- to adress this issue we need to balance the data before training model with the data.

In [110]:
imbal_ratio = 5174 / 1869
print(imbal_ratio)

2.7683253076511503


In [111]:
for col in df.columns:
    print(f"Columns: {col}")
    print(df[col].unique())

Columns: gender
['Female' 'Male']
Columns: SeniorCitizen
[0 1]
Columns: Partner
['Yes' 'No']
Columns: Dependents
['No' 'Yes']
Columns: tenure
[ 1 34  2 45  8 22 10 28 62 13 16 58 49 25 69 52 71 21 12 30 47 72 17 27
  5 46 11 70 63 43 15 60 18 66  9  3 31 50 64 56  7 42 35 48 29 65 38 68
 32 55 37 36 41  6  4 33 67 23 57 61 14 20 53 40 59 24 44 19 54 51 26  0
 39]
Columns: PhoneService
['No' 'Yes']
Columns: MultipleLines
['No phone service' 'No' 'Yes']
Columns: InternetService
['DSL' 'Fiber optic' 'No']
Columns: OnlineSecurity
['No' 'Yes' 'No internet service']
Columns: OnlineBackup
['Yes' 'No' 'No internet service']
Columns: DeviceProtection
['No' 'Yes' 'No internet service']
Columns: TechSupport
['No' 'Yes' 'No internet service']
Columns: StreamingTV
['No' 'Yes' 'No internet service']
Columns: StreamingMovies
['No' 'Yes' 'No internet service']
Columns: Contract
['Month-to-month' 'One year' 'Two year']
Columns: PaperlessBilling
['Yes' 'No']
Columns: PaymentMethod
['Electronic check' 'M

the above loop gives us unique values of each columns which helps us understand the value ranges of each. Instead of One Hot Encoding everything we will use map to handle binary values like yes:no, and for more than two values in a column we will use OHE. We could have used OHE for every object column but doing that will add up extra columns (for example : initial =  5 columns, after OHE = 10 columns), which would increase the dimension of the data as well as cost more computational speed and cost while converging. Though, we need to make sure that we only map out binary feature columns.  

Preprocessing

In [112]:
df["gender"] = df["gender"].map({"Female" : 1, "Male" : 0})
mapped = df[["Partner","Dependents","PhoneService","PaperlessBilling","Churn"]]

for col in mapped:
    df[col] = df[col].map({"Yes" : 1, "No" : 0})

In [113]:
for col in df.columns:
    print(f"Columns: {col}")
    print(df[col].unique())

Columns: gender
[1 0]
Columns: SeniorCitizen
[0 1]
Columns: Partner
[1 0]
Columns: Dependents
[0 1]
Columns: tenure
[ 1 34  2 45  8 22 10 28 62 13 16 58 49 25 69 52 71 21 12 30 47 72 17 27
  5 46 11 70 63 43 15 60 18 66  9  3 31 50 64 56  7 42 35 48 29 65 38 68
 32 55 37 36 41  6  4 33 67 23 57 61 14 20 53 40 59 24 44 19 54 51 26  0
 39]
Columns: PhoneService
[0 1]
Columns: MultipleLines
['No phone service' 'No' 'Yes']
Columns: InternetService
['DSL' 'Fiber optic' 'No']
Columns: OnlineSecurity
['No' 'Yes' 'No internet service']
Columns: OnlineBackup
['Yes' 'No' 'No internet service']
Columns: DeviceProtection
['No' 'Yes' 'No internet service']
Columns: TechSupport
['No' 'Yes' 'No internet service']
Columns: StreamingTV
['No' 'Yes' 'No internet service']
Columns: StreamingMovies
['No' 'Yes' 'No internet service']
Columns: Contract
['Month-to-month' 'One year' 'Two year']
Columns: PaperlessBilling
[1 0]
Columns: PaymentMethod
['Electronic check' 'Mailed check' 'Bank transfer (automatic)'

In [114]:
df["TotalCharges"] = pd.to_numeric(df["TotalCharges"], errors="coerce")

In [115]:
df["TotalCharges"].isna().sum()

np.int64(11)

In [116]:
df["TotalCharges"] = df["TotalCharges"].fillna(df["TotalCharges"].median())
df["TotalCharges"].isna().sum()

np.int64(0)

In [117]:
print(df["gender"])

0       1
1       0
2       0
3       0
4       1
       ..
7038    0
7039    1
7040    1
7041    0
7042    0
Name: gender, Length: 7043, dtype: int64


In [118]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 20 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   gender            7043 non-null   int64  
 1   SeniorCitizen     7043 non-null   int64  
 2   Partner           7043 non-null   int64  
 3   Dependents        7043 non-null   int64  
 4   tenure            7043 non-null   int64  
 5   PhoneService      7043 non-null   int64  
 6   MultipleLines     7043 non-null   object 
 7   InternetService   7043 non-null   object 
 8   OnlineSecurity    7043 non-null   object 
 9   OnlineBackup      7043 non-null   object 
 10  DeviceProtection  7043 non-null   object 
 11  TechSupport       7043 non-null   object 
 12  StreamingTV       7043 non-null   object 
 13  StreamingMovies   7043 non-null   object 
 14  Contract          7043 non-null   object 
 15  PaperlessBilling  7043 non-null   int64  
 16  PaymentMethod     7043 non-null   object 


In [119]:
from sklearn.model_selection import train_test_split

X = df.iloc[:,0:-1]
y = df.iloc[:,-1]

X_train, X_test, y_train, y_test = train_test_split(X,y,stratify=y,test_size=0.2,random_state=42)

In [120]:
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler

cat = ["MultipleLines","InternetService","OnlineSecurity","OnlineBackup",
          "DeviceProtection","TechSupport","StreamingTV","StreamingMovies",
          "Contract","PaymentMethod"]

num = ["tenure", "MonthlyCharges", "TotalCharges"]

preprocessor = ColumnTransformer(
    transformers=
    [("cat",OneHotEncoder(handle_unknown="ignore", sparse_output=False), cat),
     ("num",StandardScaler(), num)
     ],
    remainder="passthrough"
)

In [121]:
from sklearn.linear_model import LogisticRegression

model_logres = LogisticRegression(class_weight="balanced")

pipe_logres = Pipeline(steps=[
    ("preprocessor", preprocessor),
    ("model_logres", model_logres)
])

In [122]:
from sklearn.ensemble import RandomForestClassifier

model_rf = RandomForestClassifier(class_weight="balanced")

pipe_rf = Pipeline(steps=[
    ("preprocessor", preprocessor),
    ("model_ranfor", model_rf)
])

In [123]:
scale_pos_weight = (y_train == 0).sum() / (y_train == 1).sum()

In [124]:
from xgboost import XGBClassifier

model_xgb = XGBClassifier(
    random_state=42,
    scale_pos_weight=scale_pos_weight,
    objective="binary:logistic",
    eval_metric="logloss"
)

pipe_xgb = Pipeline(steps=[
    ("preprocess", preprocessor),
    ("model", model_xgb)
])

In [125]:
from lightgbm import LGBMClassifier

model_lgbm = LGBMClassifier(class_weight="balanced", verbose = -1)

pipe_lgbm = Pipeline(steps=[
    ("preprocessor", preprocessor),
    ("model_lgbm", model_lgbm)
])


In [126]:
from catboost import CatBoostClassifier

model_cat = CatBoostClassifier(verbose=0)

pipe_cat = Pipeline(steps=[
    ("preprocessor", preprocessor),
    ("model_cat", model_cat)
])

In [127]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score, precision_score, recall_score, f1_score

models = {
    "Logres" : pipe_logres,
    "RandomForest" : pipe_rf,
    "XGB" : pipe_xgb,
    "LightGBM" : pipe_lgbm,
    "CatBoost" : pipe_cat
}

results = []

for name, model in models.items():
    
    model.fit(X_train, y_train)
    
    y_pred = model.predict(X_test)
    
    acc = accuracy_score(y_test, y_pred)
    prec = precision_score(y_test, y_pred)
    rec = recall_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred)
    
    results.append({
        "Model": name,
        "Accuracy": acc,
        "Precision": prec,
        "Recall": rec,
        "F1 Score": f1
    })

results_df = pd.DataFrame(results)

results_df = results_df.sort_values(by="Recall", ascending=False)

print(results_df)
    


/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/utils/validation.py:2691: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


          Model  Accuracy  Precision    Recall  F1 Score
0        Logres  0.738112   0.504303  0.783422  0.613613
3      LightGBM  0.762952   0.537736  0.762032  0.630531
2           XGB  0.771469   0.556277  0.687166  0.614833
4      CatBoost  0.801987   0.656766  0.532086  0.587888
1  RandomForest  0.781405   0.615385  0.470588  0.533333


Cross Validation 

In [128]:
from sklearn.model_selection import StratifiedKFold, cross_validate

skf = StratifiedKFold(n_splits=5, shuffle=True,random_state=42)

model = {
    "Logres" : pipe_logres,
    "LightGBM" : pipe_lgbm,
    "XGBoost" : pipe_xgb
}

cv_results = {}

for name, model in models.items():
    scores = cross_validate(
        model,
        X,
        y,
        cv = skf,
        scoring=["recall", "precision", "f1"],
        return_train_score=False
    )
    
    cv_results[name] = {
        "Recall" : scores["test_recall"].mean(),
        "Precison" : scores["test_precision"].mean(),
        "F1" : scores["test_f1"].mean()
    }
    
    cv_df = pd.DataFrame(cv_results).T
    print(cv_df.sort_values(by="Recall", ascending=False)
)
    

          Recall  Precison        F1
Logres  0.801484  0.513362  0.625824
                Recall  Precison        F1
Logres        0.801484  0.513362  0.625824
RandomForest  0.481516  0.630243  0.545601
                Recall  Precison        F1
Logres        0.801484  0.513362  0.625824
XGB           0.681643  0.540315  0.602748
RandomForest  0.481516  0.630243  0.545601


/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/utils/validation.py:2691: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/utils/validation.py:2691: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/utils/validation.py:2691: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/utils/validation.py:2691: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/

                Recall  Precison        F1
Logres        0.801484  0.513362  0.625824
LightGBM      0.737822  0.528350  0.615694
XGB           0.681643  0.540315  0.602748
RandomForest  0.481516  0.630243  0.545601
                Recall  Precison        F1
Logres        0.801484  0.513362  0.625824
LightGBM      0.737822  0.528350  0.615694
XGB           0.681643  0.540315  0.602748
CatBoost      0.520033  0.650008  0.577447
RandomForest  0.481516  0.630243  0.545601


Threshold Tuning

In [129]:
models = {
    "LogReg": pipe_logres,
    "LightGBM": pipe_lgbm
}

threshold = 0.3
results = []

for name, model in models.items():
    
    y_prob_thres = model.predict_proba(X_test)[:, 1]
    y_pred_thres = (y_prob_thres > threshold).astype(int)
    
    precision = precision_score(y_test, y_pred_thres)
    recall = recall_score(y_test, y_pred_thres)
    f1 = f1_score(y_test, y_pred_thres)
    
    tn, fp, fn, tp = confusion_matrix(y_test, y_pred_thres).ravel()
    
    results.append({
        "Model": name,
        "Threshold": threshold,
        "Precision": precision,
        "Recall": recall,
        "F1 Score": f1,
        "FN (Missed Churn)": fn,
        "FP (False Alarm)": fp
    })

results_df = pd.DataFrame(results)

print(results_df)

      Model  Threshold  Precision    Recall  F1 Score  FN (Missed Churn)  \
0    LogReg        0.3   0.429455  0.927807  0.587140                 27   
1  LightGBM        0.3   0.458689  0.860963  0.598513                 52   

   FP (False Alarm)  
0               461  
1               380  


/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/utils/validation.py:2691: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


Saving the model

In [130]:
import joblib

joblib.dump(pipe_lgbm, "model.pkl")
print("Model Saved Sucessfully")

Model Saved Sucessfully
